### DISCUSSION TITLE
Template_File to templatefile migration

### DISCUSSION THREAD

**Discussion by matthewlittle** [gruntwork-io/knowledge-base#350] [Help] [answered]
I'm upgrading our usage of `template_file` (deprecated) to `templatefile` and I could use a little help with the following:

```data "template_file" "vpn_routes" {
  count    = length(data.template_file.all_vpc_cidr_blocks.*.rendered)
  template = "${cidrhost(element(data.template_file.all_vpc_cidr_blocks.*.rendered, count.index), 0)} ${cidrnetmask(element(data.template_file.all_vpc_cidr_blocks.*.rendered, count.index))}"
}

data "template_file" "all_vpc_cidr_blocks" {
  count    = length(data.terraform_remote_state.other_vpcs.*.vpc_cidr_block) + 1
  template = element(concat(data.terraform_remote_state.other_vpcs.*.vpc_cidr_block, list(data.terraform_remote_state.mgmt_vpc.vpc_cidr_block)), count.index)
}

```

These aren't standard paths to templates, so I'm not sure how to migrate them.

**Comment by rhoboat** [comment]
Hey @matthewlittle! Could you be more specific about where this template file is located, and more about the context behind this migration?

**Comment by matthewlittle** [comment]
Sure @rhoboat, happy to add more context.

Our entire dev group has upgraded to new Apple hardware w/ their new chipset. As such we're unable to manage our deployments with the current version (0.11.15). I'd like to get it up to the most recent version (and continue to keep it up to date).

Both outputs from the `template_file` above are piped into the `ecs_cluster` module as 

```
 alb_security_group_ids     = [compact(concat(data.template_file.alb_public_security_group_ids.*.rendered, data.template_file.alb_internal_security_group_ids.*.rendered))]
  num_alb_security_group_ids = [var.allow_requests_from_public_alb, var.allow_requests_from_internal_alb]
```

**Comment by yorinasub17** [comment, marked as answer]
Since you are using `template_file` as a string template rather than file, the proper replacement here is actually `locals` with plain string interpolation instead of `templatefile`:

```
locals {
  all_vpc_cidr_blocks = concat([data.terraform_remote_state.mgmt_vpc.vpc_cidr_block], data.terraform_remote_state.other_vpcs.*.vpc_cidr_block)

  all_vpc_cidr_nets = [
    for vpc_cidr in local.all_vpc_cidr_blocks :
    "${cidrhost(vpc_cidr)} ${cidrnetmask(vpc_cidr)}"
  ]
}
```

**Comment by matthewlittle** [comment]
@yorinasub17 another question related to this: 

```
data "template_file" "alb_public_security_group_ids" {
  count    = var.allow_requests_from_public_alb
  template = element(data.terraform_remote_state.alb_public.*.alb_security_group_id, count.index)
}
```

The `count` variable is actually a boolean. I'm not sure how it works with `count.index`. This code was initially setup back in 2018 and hasn't been touched since. I'm assuming it was used a way of expressing conditional logic in older versions of terraform. Any insight there?

**Reply by yorinasub17** [reply]
In terraform pre 0.12, bools were automatically translated to numbers, so it is equivalent to:

```
count = bool ? 1 : 0
```

So `element(data.terraform_remote_state.alb_public.*.alb_security_group_id, count.index)` is the same as `element(data.terraform_remote_state.alb_public.*.alb_security_group_id, 0)`
